gobblin-runtime/src/main/java/gobblin/runtime/util/FSJobCatalogHelper.java

```diff
@@ -0,0 +1,231 @@
+
+  /**
+   * Used for shadow copying in the process of updating a existing job configuration file,
+   * which requires deletion of the pre-existed copy of file and create a new one with the same name.
+   * Steps:
+   *  Create a new one in /tmp.
+   *  Safely deletion of old one.
+   *  copy the newly createad configuration file to jobConfigDir.
+   *  Delete the shadow file.
+   *
+   * @param srcFileName The target file to be replaced,
+   *               Noted that only the file name is provided, to guarantee in /tmp they are flat.
+   */
+  public static void materializedJobSpec(Path srcFileName, Path jobConfigDirPath, JobSpec jobSpec, boolean preExisted)
+      throws IOException, JobSpecNotFoundException {
+    Path shadowDirectoryPath = new Path("/tmp");
+    Configuration conf = new Configuration();
+    try (FileSystem shadowFS = shadowDirectoryPath.getFileSystem(conf)) {
+      Path shadowFilePath = PathUtils.mergePaths(shadowDirectoryPath, srcFileName);
+      System.err.println("shadow Path:" + shadowFilePath);
+      /* If previously existed, should delete anyway */
+      if (shadowFS.exists(shadowFilePath)) {
+        shadowFS.delete(shadowFilePath, false);
+      }
+
+      /*Write new file in /tmp folder*/
+      Properties props = ConfigUtils.configToProperties(jobSpec.getConfig());
+      // Persist the metadata info as well.
+      props.setProperty(DESCRIPTION_KEY_IN_JOBSPEC, jobSpec.getDescription());
+      props.setProperty(VERSION_KEY_IN_JOBSPEC, jobSpec.getVersion());
+      try (OutputStream os = shadowFS.create(shadowFilePath)) {
+        BufferedWriter bufferedWriter = new BufferedWriter(new OutputStreamWriter(os, Charsets.UTF_8));
+        props.store(bufferedWriter, "");
+        bufferedWriter.close();
+      }
+
+      /* (Optionally:Delete oldSpec) and copy the new one in. */
+      try (FileSystem configFs = jobConfigDirPath.getFileSystem(conf)) {
+        Path toBeReplacedPath = resovledRelativePath(jobConfigDirPath, new Path(jobSpec.getUri()));
+        synchronized (FSJobCatalog.class) {
+          if (preExisted) {
+            if (!configFs.exists(toBeReplacedPath)) {
+              throw new JobSpecNotFoundException(jobSpec.getUri());
+            }
+            configFs.delete(toBeReplacedPath, false);
+          }
+          FileUtil.copy(shadowFS, shadowFilePath, configFs, toBeReplacedPath, true, true, conf);
+        }
+      }
+    }
+  }
+
+  /**
+   * Instance of this function is passed as a parameter into other transform function,
+   * for converting Properties object into JobSpec object.
+   */
+  public static class JobSpecConverter implements Function<Config, JobSpec> {
+    private Path jobConfigDirPath;
+
+    public JobSpecConverter(Path jobConfigDirPath) {
+      this.jobConfigDirPath = jobConfigDirPath;
+    }
+
+    @Nullable
+    @Override
+    public JobSpec apply(@Nullable Config inputConfig) {
+      return jobSpecConverter(inputConfig, jobConfigDirPath);
```

Comment:
The function interface is only used for transforming each elements in list into the type that we expected. But this method has other usage, like in `FSJobCatalog:170` and `ImmutableFSJobCatalog:82`